harness/determined/pytorch/_pytorch_context.py

```diff
@@ -492,13 +543,31 @@ def clip_grads(params):
                 )
 
             if clip_grads is not None:
+                if self._scaler and self._auto_amp:
+                    self._scaler.unscale()
                 clip_grads(parameters)
 
-            if self.hvd_config.use:
-                with optimizer.skip_synchronize():  # type: ignore
-                    optimizer.step()
+            # For stepping the optimizer we will operate on the scaler passed
+            # in, or fall back to the wrapped scaler (if any).
+            if scaler is None:
+                scaler = self._scaler
+            if scaler:
+                if self.hvd_config.use:
+                    # TODO can we do a type assertion for horovod.torch._DistributedOptimizer
+                    with optimizer.skip_synchronize():  # type: ignore
+                        scaler.step(optimizer)
+                else:
+                    scaler.step(optimizer)
             else:
-                optimizer.step()
+                if self.hvd_config.use:
+                    # TODO can we do a type assertion for horovod.torch._DistributedOptimizer
```

Comment:
Answer: definitely not because we do not have a hard dependency on horovod.

Also, it is our responsibility to wrap the optimizer in the first place, so we should be capable of tracking that information.

Reply:
I had forgotten about this. Makes sense (avoiding the hard dependency) - but I had to add a lot of `# type: ignore`s recently and I didn't want to use them where being more specific was possible.